caffe2/proto/caffe2.proto

```diff
@@ -152,11 +202,27 @@ message DeviceOption {
 // Operator Definition.
 message OperatorDef {
   repeated string input = 1; // the name of the input blobs
+  // we plan to support named inputs in schema, and this extension is for the
+  // named inputs
+  // if inputs are positional, just leave mapped_inputs empty
+  // if (some) inputs are named in the schema, the positional inputs have
+  // empty string as place holder in mapped_inputs, and the named inputs
+  // have their schema input names at the correponding postition in
+  // mapped_inputs
+  repeated string mapped_inputs = 11;
```

Comment:
there are several things that get conflated here:
1) should we use names or positions for serialization? There's an argument that even for C2 nets it's much better to serialize names as it has clearer semantics for adding/removing inputs when the need occurs.
2) do we have distinction between inputs and arguments on serialization format or we try to enforce it downstream when op is instantiated?
3) are inputs which are constants specified as input constant nodes in the graph or are they present on the node itself.

My slight preference for the ideal world is:
1) everything is names so we  can add/remove inputs freely
2) it's enforced downstream as a way to register some kernels that require some inputs to be constants (hence arguments)
3) no strong preference for the serialization purposes

From the delta with Caffe2 - the decision we need to make here is whether we want to enable (1) and add names to all operators within C2 or do some trick with putting just a digit there.

Reply:
For item 1: I don't have a strong preference but it might be good to drill down the explicit use cases, for example how named inputs look like, because some use cases we discussed - like variable input - is actually better with a list.

For item 2 - we should enforce the distinction between inputs and arguments on serialization format. There is a clear distinction between them, in the sense that argument define the behavior of the operator (think pad in a convolution) while inputs are basically numerical contents that the operator manipulates.

For item 3 - there is a distinction between "constant input" vs "arguments". Some earlier discussions conflate them and predetermined that arguments are constant input, which is premature decision. For arguments - see above item 2. For constant input, can we drill down on use cases? One use case I can think of is constant parameters (that is very common in inference) - in these cases they should be input constant nodes.

Reply:
Yeah, I will remove this prematured design. :-)

Reply:
For 3 - I can actually see pros and cons for both versions.
Pros for args:
- It's simplifies all static type/shape inference logic. As an example "Slice" is one of the operators that at the moment have 2 versions and in many cases version with Argument is preferred for this reason.
- It simplifies setup with multiple machines, since you don't need to create blobs in multiple workspaces.
Cons:
- For inputs of large size - replication/storage will become a problem.